### ISSUE TITLE
queryMessagesWithTotal

### ISSUE THREAD

**Issue by itmagus** [turms-im/turms#1047]
https://github.com/turms-im/turms/issues/1020
queryMessagesWithTotal 的API，是否适合会话列表
https://github.com/turms-im/turms/issues/1039
turms.service.message.use-conversation-id=true.
turmsClient.messageService.queryMessages({areGroupMessages: false, fromIds: [10,11,12]}), for example.
1.目前queryMessagesWithTotal和queryMessages，在不传fromIds的时候，都只能查询到对方发送的消息
2.目前queryMessagesWithTotal在传了fromIds的情况下，最后一条消息可以是自己发送的消息
3.这样的话，事先拉取，联系人和群组列表，是否就可以实现会话列表和最后一条消息的显示呢？
4.如果是这样的话，queryMessagesWithTotal为什么在开启turms.service.message.use-conversation-id=true后，一定要传fromIds呢，才能达到查询自己发送的消息呢？queryMessages的应该场景传fromIds是应该的，进入会话后才需要消费数据

**Comment by JamesChenX** [comment]
3. Yes. You need to cache the contacts and groups on your local client just as mentioned in https://github.com/turms-im/turms/issues/1020#issuecomment-1266701025. And the data should be enough for showing contacts and messages on UI.
Notes: You need to update the cache yourself when your get change notifications (e.g. a group info change), and query the latest data with the `lastUpdatedDate` when logged in just as mentioned in the above link.

4. It is designed in this way because of its performance.  When `turms.service.message.use-conversation-id=true`:
For group conversations, the client doesn't need to pass `fromIds` (groupIds) because no matter whether the client passes the `fromIds/groupIds` or not, turms-server need to validate whether the user is in the groups (a user cannot query the messages in the groups that he/she isn't the group member) or query in which groups the user is.

For private conversations, we require the client to pass `fromIds` because 1. In a real application, the client should have already known its contacts before querying their messages, and it's reasonable. 2. With the `fromIds` passed from the client, turms-service can just send ONE request to MongoDB without the need for validation to query the messages.

**Comment by itmagus** [comment]
You can reference and understand this example, but don't must follow it. Once you understand it, you can orchestrate client APIs yourself.
嗯，我试着在理解，api应用的场景
我是这么认为的，会话列表的特点是
1.联系人，或者群组，且存在至少一条历史消息，会话列表，不等于联系人列表。
2.在用户不点击某一个会话前，不需要去拉取每会话所有的消息到本地，只需要维护好会话列表的本地数据就好了。
3.当用户进入会话后，通过进入事件、滚动、下拉等事件，按需加载消息到本地；分散用户初始化或登录时需要做的事。
4.关于消息通知，解决的应该是及时影响的问题，只是在线情况下的方式
5.如果设计是这样的，那就这么用，反正本地肯定有fromIds
6.如果有更好的API或方式来实现类似的场景，麻烦告诉我咯

**Comment by JamesChenX** [comment]
Just a reminder: The features of the conversations you mentioned are just for your application's scenarios, meaning turms or other applications don't need to follow, and that's also why 1. turms provide a lot of basic APIs for developers to compose by themselves; 2. turms clients themselves never send requests actively except heartbeat requests.

> 如果设计是这样的，那就这么用

**Turms do NOT design in this way (in fact, turms isn't just designed for any specific scenarios. For details: https://turms-im.github.io/docs/for-developers/schema.html#%E9%9C%80%E6%B1%82%E5%88%86%E6%9E%90%E4%B8%8E%E9%9B%86%E5%90%88%E7%BB%93%E6%9E%84%E8%AE%BE%E8%AE%A1) but can work efficiently with your design.** So, you need to analyze your business requirements and design your application yourself with your project managers usually, and if there are some common business requirements that you think turms should better support, we will support them.

> 如果有更好的API或方式来实现类似的场景，麻烦告诉我咯
1. All client APIs are efficient if you use them correctly: Always limit the query ranges (e.g. always passing `lastUpdatedDate` unless the device logged in the first time without any local cache).
2. All client APIs work for different features, so there aren't `better` APIs. But you can require turms to add new features if you think it's a common IM feature.

**Comment by itmagus** [comment]
嗯，好的，谢谢，我选择更好的遵循和使用Turms的设计和使用原则

**Comment by itmagus** [comment]
![image](https://user-images.githubusercontent.com/8073980/196451438-83f32a4e-c4c4-42dc-8815-db114d64897c.png)
这个API可以说明一下，参数和执行的结果是否相符？

**Comment by JamesChenX** [comment]
@itmagus No. And you are not using the latest code in `develop` or `master` branch, just update your code and it's will be fixed.

**Comment by lovemianhuatang** [comment]
记录：默认是maxCount=1按deliveryDate升序，所以如果要获取最新一条消息需要增加descending: true参数
